## Start with cProfilling

In [1]:
import cProfile
import pstats

def sample_function():
    # Some code here to profile
    sum = 0
    for i in range(10000):
        sum += i
    return sum

# Profiling the function
cProfile.run('sample_function()', 'profile_stats')

# Reading the stats
p = pstats.Stats('profile_stats')
p.sort_stats('cumulative').print_stats(10)  # prints the top 10 functions where time was spent


Sun Dec 10 18:25:21 2023    profile_stats

         4 function calls in 0.000 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 /tmp/ipykernel_63955/745881689.py:4(sample_function)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [3]:
# example.py
def compute():
    data = [i for i in range(1000000)]
    sum = 0
    for i in data:
        sum += i
    return sum

if __name__ == "__main__":
    compute()


In [2]:
!valgrind --tool=cachegrind python compute.py

==32138== Cachegrind, a high-precision tracing profiler
==32138== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
==32138== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==32138== Command: python compute.py
==32138== 
==32138== brk segment overflow in thread #1: can't grow to 0x4aae000
==32138== (see section Limitations in user manual)
==32138== NOTE: further instances of this message will not be shown
==32138== 
==32138== I refs:        555,382,806


> *I refs:        555,382,806 --> It's a count of how many instructions were fetched from memory.*

In [4]:
# !valgrind --tool=cachegrind python cache_miss_example.py
!valgrind --tool=cachegrind --cache-sim=yes python NeighboorhoodAggregation.py

==64814== Cachegrind, a cache and branch-prediction profiler
==64814== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
==64814== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==64814== Command: python NeighboorhoodAggregation.py
==64814== 
--64814-- warning: L3 cache found, using its data for the LL simulation.
--64814-- WARNING: Serious error when reading debug info
--64814-- When reading debug info from /lustre/acslab/users/3281/miniconda3/envs/cachemisses/bin/python3.11:
--64814-- Ignoring non-Dwarf2/3/4 block in .debug_info
--64814-- WARNING: Serious error when reading debug info
--64814-- When reading debug info from /lustre/acslab/users/3281/miniconda3/envs/cachemisses/bin/python3.11:
--64814-- Ignoring non-Dwarf2/3/4 block in .debug_info
--64814-- WARNING: Serious error when reading debug info
--64814-- When reading debug info from /lustre/acslab/users/3281/miniconda3/envs/cachemisses/bin/python3.11:
--64814-- Ignoring non-Dwarf2/3/4 bl

In [6]:
!valgrind --tool=cachegrind --cache-sim=yes python cache_miss_example.py

==16224== Cachegrind, a cache and branch-prediction profiler
==16224== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
==16224== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==16224== Command: python cache_miss_example.py
==16224== 
--16224-- warning: L3 cache found, using its data for the LL simulation.
--16224-- WARNING: Serious error when reading debug info
--16224-- When reading debug info from /lustre/acslab/users/3281/miniconda3/envs/cachemisses/bin/python3.11:
--16224-- Ignoring non-Dwarf2/3/4 block in .debug_info
--16224-- WARNING: Serious error when reading debug info
--16224-- When reading debug info from /lustre/acslab/users/3281/miniconda3/envs/cachemisses/bin/python3.11:
--16224-- Ignoring non-Dwarf2/3/4 block in .debug_info
--16224-- WARNING: Serious error when reading debug info
--16224-- When reading debug info from /lustre/acslab/users/3281/miniconda3/envs/cachemisses/bin/python3.11:
--16224-- Ignoring non-Dwarf2/3/4 block in